File: RadarSDK/RadarTripOrder.m
//
//  RadarTripOrder.m
//  RadarSDK
//
//  Copyright © 2024 Radar Labs, Inc. All rights reserved.
//

#import "RadarTripOrder.h"
#import "RadarUtils.h"

@implementation RadarTripOrder

- (instancetype _Nullable)initWithId:(NSString *_Nonnull)_id
                                guid:(NSString *_Nullable)guid
                        handoffMode:(NSString *_Nullable)handoffMode
                              status:(RadarTripOrderStatus)status
                             firedAt:(NSDate *_Nullable)firedAt
                       firedAttempts:(NSNumber *_Nullable)firedAttempts
                         firedReason:(NSString *_Nullable)firedReason
                           updatedAt:(NSDate *_Nonnull)updatedAt {
    self = [super init];
    if (self) {
        __id = _id;
        _guid = guid;
        _handoffMode = handoffMode;
        _status = status;
        _firedAt = firedAt;
        _firedAttempts = firedAttempts;
        _firedReason = firedReason;
        _updatedAt = updatedAt;
    }
    return self;
}

- (instancetype _Nullable)initWithObject:(NSObject *)object {
    if (!object || ![object isKindOfClass:[NSDictionary class]]) {
        return nil;
    }

    NSDictionary *dict = (NSDictionary *)object;

    NSString *_id;
    NSString *guid;
    NSString *handoffMode;
    RadarTripOrderStatus status = RadarTripOrderStatusUnknown;
    NSDate *firedAt;
    NSNumber *firedAttempts;
    NSString *firedReason;
    NSDate *updatedAt;

    id idObj = dict[@"id"];
    if (idObj && [idObj isKindOfClass:[NSString class]]) {
        _id = (NSString *)idObj;
    }

    id guidObj = dict[@"guid"];
    if (guidObj && [guidObj isKindOfClass:[NSString class]]) {
        guid = (NSString *)guidObj;
    }

    id handoffModeObj = dict[@"handoffMode"];
    if (handoffModeObj && [handoffModeObj isKindOfClass:[NSString class]]) {
        handoffMode = (NSString *)handoffModeObj;
    }

    id statusObj = dict[@"status"];
    if (statusObj && [statusObj isKindOfClass:[NSString class]]) {
        NSString *statusStr = (NSString *)statusObj;
        if ([statusStr isEqualToString:@"pending"]) {
            status = RadarTripOrderStatusPending;
        } else if ([statusStr isEqualToString:@"fired"]) {
            status = RadarTripOrderStatusFired;
        } else if ([statusStr isEqualToString:@"canceled"]) {
            status = RadarTripOrderStatusCanceled;
        } else if ([statusStr isEqualToString:@"completed"]) {
            status = RadarTripOrderStatusCompleted;
        }
    }

    id firedAtObj = dict[@"firedAt"];
    if (firedAtObj && [firedAtObj isKindOfClass:[NSString class]]) {
        NSString *firedAtStr = (NSString *)firedAtObj;
        firedAt = [RadarUtils.isoDateFormatter dateFromString:firedAtStr];
    }

    id firedAttemptsObj = dict[@"firedAttempts"];
    if (firedAttemptsObj && [firedAttemptsObj isKindOfClass:[NSNumber class]]) {
        firedAttempts = (NSNumber *)firedAttemptsObj;
    }

    id firedReasonObj = dict[@"firedReason"];
    if (firedReasonObj && [firedReasonObj isKindOfClass:[NSString class]]) {
        firedReason = (NSString *)firedReasonObj;
    }

    id updatedAtObj = dict[@"updatedAt"];
    if (updatedAtObj && [updatedAtObj isKindOfClass:[NSString class]]) {
        NSString *updatedAtStr = (NSString *)updatedAtObj;
        updatedAt = [RadarUtils.isoDateFormatter dateFromString:updatedAtStr];
    }

    if (_id && updatedAt) {
        return [[RadarTripOrder alloc] initWithId:_id
                                             guid:guid
                                     handoffMode:handoffMode
                                           status:status
                                          firedAt:firedAt
                                    firedAttempts:firedAttempts
                                      firedReason:firedReason
                                        updatedAt:updatedAt];
    }

    return nil;
}

+ (NSArray<RadarTripOrder *> *_Nullable)ordersFromObject:(id _Nonnull)object {
    if (!object || ![object isKindOfClass:[NSArray class]]) {
        return nil;
    }

    NSArray *ordersArr = (NSArray *)object;
    NSMutableArray<RadarTripOrder *> *mutableOrders = [NSMutableArray<RadarTripOrder *> new];

    for (id orderObj in ordersArr) {
        RadarTripOrder *order = [[RadarTripOrder alloc] initWithObject:orderObj];
        if (!order) {
            return nil;
        }
        [mutableOrders addObject:order];
    }

    return mutableOrders;
}

+ (NSArray<NSDictionary *> *)arrayForOrders:(NSArray<RadarTripOrder *> *)orders {
    if (!orders) {
        return nil;
    }

    NSMutableArray *arr = [[NSMutableArray alloc] initWithCapacity:orders.count];
    for (RadarTripOrder *order in orders) {
        NSDictionary *dict = [order dictionaryValue];
        [arr addObject:dict];
    }
    return arr;
}

+ (NSString *)stringForStatus:(RadarTripOrderStatus)status {
    switch (status) {
    case RadarTripOrderStatusPending:
        return @"pending";
    case RadarTripOrderStatusFired:
        return @"fired";
    case RadarTripOrderStatusCanceled:
        return @"canceled";
    case RadarTripOrderStatusCompleted:
        return @"completed";
    default:
        return @"unknown";
    }
}

- (NSDictionary *)dictionaryValue {
    NSMutableDictionary *dict = [NSMutableDictionary new];
    [dict setValue:self._id forKey:@"id"];
    [dict setValue:self.guid forKey:@"guid"];
    [dict setValue:self.handoffMode forKey:@"handoffMode"];
    [dict setValue:[RadarTripOrder stringForStatus:self.status] forKey:@"status"];
    if (self.firedAt) {
        NSString *firedAtString = [RadarUtils.isoDateFormatter stringFromDate:self.firedAt];
        [dict setValue:firedAtString forKey:@"firedAt"];
    }
    [dict setValue:self.firedAttempts forKey:@"firedAttempts"];
    [dict setValue:self.firedReason forKey:@"firedReason"];
    NSString *updatedAtString = [RadarUtils.isoDateFormatter stringFromDate:self.updatedAt];
    [dict setValue:updatedAtString forKey:@"updatedAt"];
    return dict;
}

@end 

File: RadarSDK/RadarTripOrder.swift
//
//  RadarTripOrder.swift
//  RadarSDK
//
//  Created by Alan Charles on 8/31/26.
//  Copyright © 2026 Radar Labs, Inc. All rights reserved.
//

import Foundation

/// Keep one initialized value so failed Objective-C initializers can deallocate safely.
private final class RadarTripOrderStorage {
    let id: String
    let guid: String?
    let handoffMode: String?
    let status: RadarTripOrderStatus
    let firedAt: Date?
    let firedAttempts: NSNumber?
    let firedReason: String?
    let updatedAt: Date

    init(
        id: String,
        guid: String?,
        handoffMode: String?,
        status: RadarTripOrderStatus,
        firedAt: Date?,
        firedAttempts: NSNumber?,
        firedReason: String?,
        updatedAt: Date
    ) {
        self.id = id
        self.guid = guid
        self.handoffMode = handoffMode
        self.status = status
        self.firedAt = firedAt
        self.firedAttempts = firedAttempts
        self.firedReason = firedReason
        self.updatedAt = updatedAt
    }
}

@objc @implementation extension RadarTripOrder {

    @nonobjc private var storage: RadarTripOrderStorage?

    // swiftlint:disable:next identifier_name
    public var _id: String! {
        storage?.id
    }
    public var guid: String? {
        storage?.guid
    }
    public var handoffMode: String? {
        storage?.handoffMode
    }
    public var status: RadarTripOrderStatus {
        storage?.status ?? .unknown
    }
    public var firedAt: Date? {
        storage?.firedAt
    }
    public var firedAttempts: NSNumber? {
        storage?.firedAttempts
    }
    public var firedReason: String? {
        storage?.firedReason
    }
    public var updatedAt: Date {
        storage?.updatedAt ?? Date(timeIntervalSince1970: 0)
    }

    public init?(
        id: String,
        guid: String?,
        handoffMode: String?,
        status: RadarTripOrderStatus,
        firedAt: Date?,
        firedAttempts: NSNumber?,
        firedReason: String?,
        updatedAt: Date
    ) {
        self.storage = RadarTripOrderStorage(
            id: id,
            guid: guid,
            handoffMode: handoffMode,
            status: status,
            firedAt: firedAt,
            firedAttempts: firedAttempts,
            firedReason: firedReason,
            updatedAt: updatedAt
        )

        super.init()
    }

    @objc(initWithObject:)
    public convenience init?(object: Any) {
        guard let dictionary = object as? [AnyHashable: Any],
            let id = dictionary["id"] as? String,
            let updatedAtString = dictionary["updatedAt"] as? String,
            let updatedAt = RadarUtils.isoDateFormatter.date(
                from: updatedAtString
            )
        else {
            return nil
        }

        let guid = dictionary["guid"] as? String
        let handoffMode = dictionary["handoffMode"] as? String
        let status = Self.status(
            from: dictionary["status"] as? String
        )

        let firedAt = (dictionary["firedAt"] as? String).flatMap {
            RadarUtils.isoDateFormatter.date(from: $0)
        }

        let firedAttempts = dictionary["firedAttempts"] as? NSNumber
        let firedReason = dictionary["firedReason"] as? String

        self.init(
            id: id,
            guid: guid,
            handoffMode: handoffMode,
            status: status,
            firedAt: firedAt,
            firedAttempts: firedAttempts,
            firedReason: firedReason,
            updatedAt: updatedAt
        )
    }

    @objc(ordersFromObject:)
    public class func orders(
        from object: Any
    ) -> [RadarTripOrder]? {
        guard let objects = object as? [Any] else {
            return nil
        }

        var orders: [RadarTripOrder] = []
        orders.reserveCapacity(objects.count)

        for object in objects {
            guard let order = RadarTripOrder(object: object) else {
                return nil
            }

            orders.append(order)
        }

        return orders
    }

    @objc(arrayForOrders:)
    public class func array(
        for orders: [RadarTripOrder]?
    ) -> [[AnyHashable: Any]]? {
        orders?.map { $0.dictionaryValue() }
    }

    public func dictionaryValue() -> [AnyHashable: Any] {
        var dictionary: [AnyHashable: Any] = [
            "id": _id,
            "status": Self.string(for: status),
            "updatedAt": RadarUtils.isoDateFormatter.string(
                from: updatedAt
            ),
        ]

        if let guid {
            dictionary["guid"] = guid
        }

        if let handoffMode {
            dictionary["handoffMode"] = handoffMode
        }

        if let firedAt {
            dictionary["firedAt"] =
                RadarUtils.isoDateFormatter.string(from: firedAt)
        }

        if let firedAttempts {
            dictionary["firedAttempts"] = firedAttempts
        }

        if let firedReason {
            dictionary["firedReason"] = firedReason
        }

        return dictionary
    }

    private static func status(
        from string: String?
    ) -> RadarTripOrderStatus {
        switch string {
        case "pending":
            return .pending
        case "fired":
            return .fired
        case "canceled":
            return .canceled
        case "completed":
            return .completed
        default:
            return .unknown
        }
    }

    private static func string(
        for status: RadarTripOrderStatus
    ) -> String {
        switch status {
        case .pending:
            return "pending"
        case .fired:
            return "fired"
        case .canceled:
            return "canceled"
        case .completed:
            return "completed"
        default:
            return "unknown"
        }
    }
}
